Bolt tslint to eslint

kian2attari opened the pull request:
###  Summary

This PR aims to migrate Bolt.js from TSLint to ESLint. I used the **`typescript-eslint`** plugin to allow ESLint to work with TypeScript. I decided to go with this plugin as it was supported by both the backers behind TSLint (Palantir),  who   [deprecated TSLint in favor of   typescript-eslint](https://medium.com/palantir/tslint-in-2019-1a144c2317a9), and the actual [TypeScript team themselves](https://github.com/microsoft/TypeScript/issues/30553).

### Details

In the process of migrating Bolt to ESLint, I used the [tslint-to-eslint-config](https://github.com/typescript-eslint/tslint-to-eslint-config) project (as recommended by **`typescript-eslint`**) to automate the initial bulkwork of the process such as converting all the TSLint rule flag comments in all files to ESLint-compatible comments. 

To match the original TSLint configuration as closely as possible (which extended the tslint-airbnb-config), I extended the [`eslint-config-airbnb-typescript`](https://www.npmjs.com/package/eslint-config-airbnb-typescript) plugin in the .eslintrc.js file. I also included recommended plugins like @typescript-eslint/recommended-requiring-type-checking to match the TSLint behavior. I commented out/disabled some of the recommended ESLint plugins as they introduce new rules that the project currently does not follow, raising linting errors. One challenge with this migration is that ESLint is more strict than TSLint. Since the new rules would require some changes in Bolt's source code (many of which can be fixed automatically), *I'd like to discuss potentially including them as they do seem valuable.* Alongside these potential additions, I also added some JSDoc rules that are currently commented out.


I'd also like to discuss the idea of using **Prettier** for code formatting (ex. checking indentation) rather than the linter as recommended by the typescript-eslint and tslint-to-eslint-config team. Certain TSLint rules, such as "whitespace", are supposed to be replaced by Prettier when migrating to ESLint (as recommended in the [roadmap](https://github.com/typescript-eslint/typescript-eslint/blob/96342e328652e32f44f2e87b62e3d5b4f6dbd73f/packages/eslint-plugin/ROADMAP.md)) since there is no ESLint equivalent. Prettier also does a much better job at checking formatting than ESLint with regards to checking indentation ([relevant issue here](https://github.com/typescript-eslint/typescript-eslint/issues/1824)). The indent rule raises quite a few arbitrary errors which is why I decided to comment it out for this PR.

I also added src/middleware/builtin.ts to the .eslintignore file because it was the only file that was raising linting errors. Its formatting seems very different from the other files in the project and the errors are almost all formatting issues (ex: line length of 122 instead of 120, missing return types on functions, missing line breaks, not using dot notation etc). I think this is worth discussing



### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/slackapi/bolt-js?pullRequest=529) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/slackapi/bolt-js?pullRequest=529) before we can accept your contribution.<br/><hr/>**Kian Attari** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/slackapi/bolt-js?pullRequest=529) it.</sub>

seratch commented on package.json:
```diff
@@ -69,7 +76,6 @@
     "tslint": "^5.15.0",
-    "tslint-config-airbnb": "^5.11.1",
-    "typescript": "^3.7.2"
+    "typescript": "^3.9.5"
```
>The latest major version of @slack/bolt is supported to build against a minimum TypeScript version of v3.7.

We're saying that this library is built against 3.7+ in the document. https://slack.dev/bolt-js/tutorial/using-typescript

Is it possible to hold off upgrading the version?

kian2attari replied:
Yes, it's not necessary to upgrade the typescript version. I reverted it back to the original in this [commit](https://github.com/slackapi/bolt-js/pull/529/commits/0d538222344c41afa40ccf010e764e514a366354).

kian2attari commented:
###  Summary
Added Prettier to the linting process and used the recommended [prettier/@typescript-eslint]([https://github.com/prettier/eslint-config-prettier](https://github.com/prettier/eslint-config-prettier)) plugin to disable ESLint rules that conflict with Prettier. Also made some pretty major improvements to the ESLint configuration structure to make it more maintainable and readable. 

### Details

1. Incorporated Prettier into the linting process, included a basic prettier config, and disabled ESLint rules that conflict with Prettier using [prettier/@typescript-eslint]([https://github.com/prettier/eslint-config-prettier](https://github.com/prettier/eslint-config-prettier))
2. Removed the enormous bunch of unnecessary rule specifications I had in the previous version of the .eslintrc.js. The eslintrc.js (which lints all files except spec/test files) and .eslintrc.test.js files now extend from a base ESLint configuration found in the eslint-config-base folder. The `rules` property in .eslintrc.js and .eslintrc.test.js respectively is now only used to disable a couple of the strict new rules (ex. from airbnb's TypeScript ESLint config) that the project currently does not follow. Most of these are easy/automated fixes so I think we should conform to them. 
3. Changed the .eslintignore to only ignore the node_modules directory as that's more typical convention. I was previously also globally ignoring test/spec files (ex *.spec.ts) here so that the regular .eslintrc.js would not lint them, while using the --no-ignore flag for the test-lint command so that the test files could be linted there. Instead of global ignores, I'm now locally ignoring the test files using the `ignorePatterns`  property in .eslintrc.js.  
4. I updated some of the old TSLint line-specific rule disable comments to their ESLint counterparts and also added a few ones in new places since ESLint is so picky. I made sure not to change any of the actual code in the project files.
5. Modified some of the ESLint rule configurations to match the rules of the old tslint configuration we had even more closely
6. Updated the linting commands in package.json so be more straightforward and conventional to ESLint. Rather than specifying a glob  like `src/**/*/*.ts` like we did with the TSLint commands, I specify the file extension(s) and the directory that should be linted.  
7. I removed a couple of now unused linting-related packages and deleted tslint.json :sweep:

stevengill commented on .eslintrc.js:
```diff
@@ -0,0 +1,63 @@
+       of action is likely to adopt these rules and make the quick (and mostly automated) fixes
+       needed in the repo to conform to these. ESLint and the airbnb-typecript config is more strict
+       than the original TSLint configuration that this project had. */
+		'import/first': ['off'],
```
Definitely might want to conform to some of these airbnb rules.

stevengill commented:
I did a few updates to this branch.

1) I've setup `prettier` to actually run on save in `vscode` and added it to the `npm run lint` command.
2) I removed some of the rules in `.eslintrc.js` that were put in place to work with our existing linting but are different than the standard airbnb typescript lint rules. I kept a few as it wasn't worth trying to refactor the code at the moment to make them pass. This means this PR now includes some style changes in our code. Take a look and decide if the changes that have landed are acceptable. They were all pretty minor to me.
